types/comparable_bytes.cc

```diff
@@ -592,11 +600,88 @@ struct to_comparable_bytes_visitor {
-
+    // Encode variable length natively byte-comparable data types.
+    // Zeros in the source are escaped as ESCAPE followed by one or more ESCAPED_0_CONT and ending with ESCAPED_0_DONE.
+    // The encoding terminates in an escaped state (either with ESCAPE or ESCAPED_0_CONT).
+    // If the source ends in a zero, we use ESCAPED_0_CONT to ensure that the encoding remains smaller than
+    // the same source with an additional zero at the end.
+    //
+    // E.g. "A\0\0B" translates to 4100FEFF4200
+    //      "A\0B\0"               4100FF4200FE
+    //      "A\0"                  4100FE
+    //      "AB"                   414200
+    //
+    // In a single-byte source, the bytes could be passed unchanged, but this would prevent combining components.
+    // This translation preserves order, and since the encoding for 0 is higher than the separator,
+    // it also ensures that shorter components are treated as smaller.
+    //
+    // The encoding is not prefix-free since, for example, the encoding of "A" (4100) is a prefix of the encoding of "A\0" (4100FE).
+    // However, the byte following the prefix is guaranteed to be FE or FF, making the encoding weakly prefix-free.
+    // Additionally, any such prefix sequence will compare as smaller than the value it prefixes,
+    // because any permitted separator byte will be smaller than the byte following the prefix.
     void operator()(const abstract_type& type) {
-        // Unimplemented
-        on_internal_error(cblogger, fmt::format("byte comparable format not supported for type {}", type.name()));
+        if (!type.is_byte_order_equal()) {
+            on_internal_error(cblogger, fmt::format("byte comparable format not supported for type {}", type.name()));
+        }
+
+        bool escaped = false;
+        while (!serialized_bytes_view.empty()) {
+            const auto& curr_frag = serialized_bytes_view.current_fragment();
+            serialized_bytes_view.remove_current();
+            auto curr_zero_start = curr_frag.find(ESCAPE);
+            if (curr_zero_start == managed_bytes_view::fragment_type::npos) {
+                // There are no zeros in current fragment
+                if (escaped) {
+                    // but the previous fragment ended with a zero, so end the escape sequence
+                    write_native_int(out, ESCAPED_0_DONE);
+                    escaped = false;
+                }
+                // write the current fragment as it is; this is the fast path for the common text cases
+                out.write(curr_frag);
+            } else {
+                // There are zeros that need to be escaped;
+                // Use a buffer to write the encoded value. The encoded bytes will have one extra byte for every
+                // zero sequence in the source. So, worstcase, it can only be 1.5 times the size of the source.
+                std::vector<signed char> buffer(size_t(std::ceil(double(curr_frag.size()) * 1.5)));
```

Comment:
We might avoid the allocation by limiting the fragment processing size to a known value.

But at least use utils::small_vector (with an internal buffer of 1k) to avoid tiny allocations.